src/OpenTelemetry.Exporter.Jaeger/JaegerExporter.cs

```diff
@@ -113,26 +143,120 @@ internal void ApplyLibraryResource(Resource libraryResource)
         }
 
-        /// <inheritdoc/>
-        protected override void OnShutdown(int timeoutMilliseconds)
-        {
-            _ = this.JaegerAgentUdpBatcher.FlushAsync(default).GetAwaiter().GetResult();
-        }
-
         /// <inheritdoc/>
         protected override void Dispose(bool disposing)
         {
             if (!this.disposedValue)
             {
                 if (disposing)
                 {
-                    this.JaegerAgentUdpBatcher.Dispose();
+                    this.thriftClient.Dispose();
+                    this.clientTransport.Dispose();
+                    this.memoryTransport.Dispose();
+                    this.memoryProtocol.Dispose();
                 }
 
                 this.disposedValue = true;
             }
 
             base.Dispose(disposing);
         }
+
+        private void AppendSpan(JaegerSpan jaegerSpan)
+        {
+            if (this.processCache == null)
+            {
+                this.Process.Message = this.BuildThriftMessage(this.Process).ToArray();
+                this.processCache = new Dictionary<string, Process>
+                {
+                    [this.Process.ServiceName] = this.Process,
+                };
+            }
+
+            var spanServiceName = jaegerSpan.PeerServiceName ?? this.Process.ServiceName;
+
+            if (!this.processCache.TryGetValue(spanServiceName, out var spanProcess))
+            {
+                spanProcess = new Process(spanServiceName, this.Process.Tags);
+                spanProcess.Message = this.BuildThriftMessage(spanProcess).ToArray();
+                this.processCache.Add(spanServiceName, spanProcess);
+            }
+
+            var spanMessage = this.BuildThriftMessage(jaegerSpan);
+
+            jaegerSpan.Return();
+
+            var spanTotalBytesNeeded = spanMessage.Count;
+
+            if (!this.CurrentBatches.TryGetValue(spanServiceName, out var spanBatch))
+            {
+                spanBatch = new Batch(spanProcess);
+                this.CurrentBatches.Add(spanServiceName, spanBatch);
+
+                spanTotalBytesNeeded += spanProcess.Message.Length;
+            }
+
+            if (this.batchByteSize + spanTotalBytesNeeded >= this.maxPacketSize)
+            {
+                this.SendCurrentBatches();
+
+                // Flushing effectively erases the spanBatch we were working on, so we have to rebuild it.
+                spanBatch.Clear();
+                spanTotalBytesNeeded = spanMessage.Count + spanProcess.Message.Length;
+                this.CurrentBatches.Add(spanServiceName, spanBatch);
+            }
+
+            spanBatch.Add(spanMessage);
+            this.batchByteSize += spanTotalBytesNeeded;
+        }
+
+        private void SendCurrentBatches()
+        {
+            try
+            {
+                foreach (var batch in this.CurrentBatches)
+                {
+                    var task = this.thriftClient.WriteBatchAsync(batch.Value, CancellationToken.None);
```

Comment:
Just because I know someone will ask 😄 

The `thriftClient` instance Jaeger is using is hooked up to a UdpClient. The API is all async but ultimately writes to a socket buffer and returns completely synchronously. To save some perf I skipped awaiting the task. The `#if DEBUG` checks in here are just in case that ever changes.